src/axl.c

```diff
@@ -57,14 +58,83 @@ Global Variables
  * before transferring files */
 int axl_make_directories;
 
-/* AXL's flush file, SCR has one as well */
-char* axl_flush_file = NULL;
+/*
+ * Array for all the AXL_Create'd kvtree pointers.  It's indexed by the AXL id.
+ *
+ * Note: We only expand this array, we never shrink it.  This is fine since
+ * the user is only going to be calling AXL_Create() a handful of times.  It
+ * also simplifies the code if we never shrink it, and the extra memory usage
+ * is negligible, if any at all.
+ */
+kvtree **axl_kvtrees;
+static unsigned int axl_kvtrees_count = 0;
+
+static pthread_mutex_t id_lock = PTHREAD_MUTEX_INITIALIZER;
+
+/*
+ * Allocate a new kvtree and return the AXL ID for it.  If state_file is
+ * specified, then populate the kvtree with it's data.
+ */
+static int axl_alloc_id(const char* state_file)
+{
+    kvtree *new;
+
+    new = kvtree_new();
+    if (state_file) {
+         /* initialize kvtree values from state_file if we have one */
+        kvtree_read_file(state_file, new);
+    } else {
+        /*
+         * No state file, just put in an empty string since you can't use
+         * a NULL for a string in kvtree.
+         */
+        state_file = "";
+    }
+
+    kvtree_util_set_str(new, AXL_KEY_STATE_FILE, state_file);
+
+    pthread_mutex_lock(&id_lock);
+    axl_kvtrees = realloc(axl_kvtrees, sizeof(struct kvtree *) * (axl_kvtrees_count + 1));
+    axl_kvtrees[axl_kvtrees_count] = new;
+    axl_kvtrees_count++;
+    pthread_mutex_unlock(&id_lock);
+    return axl_kvtrees_count - 1;
+}
+
+/* Remove the state file for an id, if one exists */
+static void axl_remove_state_file(int id)
+{
+    kvtree *file_list = axl_kvtrees[id];
+    char *state_file = NULL;
+    kvtree_util_get_str(file_list, AXL_KEY_STATE_FILE, &state_file);
+    if (state_file && strcmp(state_file, "") != 0) {
+        unlink(state_file);
+    }
+}
 
-/* Transfer handle unique IDs */
-static int axl_next_handle_UID = -1;
+/* Remove the ID, we don't need it anymore */
+static void axl_free_id(int id)
+{
+    axl_remove_state_file(id);
+    pthread_mutex_lock(&id_lock);
+    kvtree_delete(&axl_kvtrees[id]);
+    axl_kvtrees[id] = NULL;
+    pthread_mutex_unlock(&id_lock);
+}
 
-/* tracks list of files written with transfer */
-kvtree* axl_file_lists = NULL;
+/*
+ * If the user specified a state_file then write our kvtree to it. If not, then
+ * do nothing.
+ */
+static void axl_write_state_file(int id)
+{
+    kvtree *file_list = axl_kvtrees[id];
```

Comment:
To match existing style, please move the stars next to the type rather than the variable name.
```suggestion
    kvtree* file_list = axl_kvtrees[id];
```